### PR TITLE
Fix slow extension sync from remote storage on startup

### PR DESCRIPTION
**Pull request by dstockton**
Extension sync awaited the remote stat and read calls one file at a time inside the list loop, so only the body copy was concurrent. Projects with hundreds of extension files took 10+ minutes to boot (see CMS-2896 for Slack context).

This moves the whole per-file sync (metadata compare, mkdir, read, pipeline) into the p-queue task so downloads run concurrently. Concurrency is bounded by a new EXTENSIONS_STORAGE_MAX_CONCURRENCY env var (default 20), shared by the extension sync and the marketplace install/uninstall storage queues. Invalid values log a warning and fall back to the default. On the first failure the sync stops: the error is recorded, queued work is cleared, the listing loop exits, and the error is rethrown after in-flight tasks drain, so the sync lock is never released while files are still being written.

Measured with the real syncExtensions against minio behind a 25ms latency proxy, 600 files:
- fresh sync: 34.8s before, 2.3s after (0.7s with concurrency 100)
- warm restart (metadata compare per file): 34.3s before, 2.0s after

Docs: directus/docs#789

Fixes CMS-2896

### CODE REVIEW

File: api/src/extensions/lib/sync/sync.ts [hunk 1]
```diff
@@ -71,37 +71,52 @@ export async function syncExtensions(options?: ExtensionSyncOptions): Promise<vo
...
 
-			// Ensure that the directory path exists
-			await mkdir(dirname(destinationPath), { recursive: true });
-
-			// write remote file to the local filesystem
-			const readStream = await disk.read(filepath);
-			const writeStream = createWriteStream(destinationPath);
-			queue.add(() => pipeline(readStream, writeStream));
+			// wait for all file syncs to finish, surfacing any errors
+			await Promise.all(syncTasks);
```
Comment (ComfortablyCoding) [inline api/src/extensions/lib/sync/sync.ts]: This wont stop any "already queued" tasks from still being processed in the queue. It also still "runs" the remaining promises after rejection as one cannot cancel a promise. The idle in the finally will wait for the queue to also finish but that means we are running remaining tasks when we already know a failure exists and should stop.

File: api/src/extensions/lib/sync/sync.ts [hunk 2]
```diff
@@ -71,37 +71,52 @@ export async function syncExtensions(options?: ExtensionSyncOptions): Promise<vo
...
+					// write remote file to the local filesystem
+					const readStream = await disk.read(filepath);
+					const writeStream = createWriteStream(destinationPath);
+					await pipeline(readStream, writeStream);
+				};
+
+				// Queue the whole per-file sync so the remote stat/read round trips run concurrently
+				const task = queue.add(syncFile);
+				// Attach a no-op handler so an early failure isn't an unhandled rejection while listing continues
+				task.catch(() => {});
```
Comment (ComfortablyCoding) [inline api/src/extensions/lib/sync/sync.ts]: On failure we should be stopping the sync and throwing, no reason to continue as partial syncs are invalid state. This catch looks to be due to symptom of https://github.com/directus/directus/pull/27989/changes#r3666609442.

**Review by ComfortablyCoding — CHANGES_REQUESTED** [review]
First pass review

File: api/src/extensions/lib/sync/sync.ts (hunk 1 above)
Comment (dstockton) [reply api/src/extensions/lib/sync/sync.ts]: Good catch, reworked in b5cd7d8: first failure now records the error and clears the queue, so nothing queued starts after it. The clear happens inside the task's own catch because p-queue dequeues the next job synchronously before an external catch on add() would run. In-flight promises still run to completion (can't cancel them), but they're bounded by the concurrency and we drain them before releasing the sync lock so peers don't read a half-written cache. Added a 1001-file test that fails if the clear is removed or moved back outside the task.

File: api/src/extensions/lib/sync/sync.ts (hunk 2 above)
Comment (dstockton) [reply api/src/extensions/lib/sync/sync.ts]: Agreed, the no-op catch is gone with the rework. Task errors are recorded inside the task and rethrown after the in-flight drain, so the sync stops and throws on first failure.

File: api/src/extensions/lib/sync/sync.ts [hunk 3]
```diff
@@ -71,37 +71,60 @@ export async function syncExtensions(options?: ExtensionSyncOptions): Promise<vo
 
 		// Make sure we don't overload the file handles
 		const queue = new Queue({ concurrency: 1000 });
```
Comment (ComfortablyCoding) [inline api/src/extensions/lib/sync/sync.ts]: Since we are now queueing requests to the API instead of processing them sequentially, we should review and adjust the concurrency limits to ensure we don't overwhelm the respective driver's API. Previously, the queue was only responsible for piping operations, where this was not a concern.

**Review by ComfortablyCoding — CHANGES_REQUESTED** [review]

File: api/src/extensions/lib/sync/sync.ts (hunk 3 above)
Comment (dstockton) [reply api/src/extensions/lib/sync/sync.ts]: Fair point. Replaced the hard-coded 1000 with a new EXTENSIONS_SYNC_MAX_CONCURRENCY env var in 5832229 (registered in DIRECTUS_VARIABLES and TYPE_MAP, so _FILE and casting work). Default is 20, following IMPORT_MAX_CONCURRENCY. In my minio test rig (600 files, 50ms RTT) that's 2.3s vs 34.8s on main, and bumping it to 100 gets 0.7s, so the default stays well clear of driver limits (S3 socket pool is 500) while operators with big extension sets can turn it up.

**Comment by dstockton** [comment]
Docs PR for the new env var: directus/docs#789

**Comment by dstockton** [comment]
CI note: the two mssql blackbox jobs hit the 1h timeout and get cancelled. Same on recently merged PRs (#27994, #27992), so it's a pre-existing infra issue, not this change (blackbox doesn't set EXTENSIONS_LOCATION, so the sync path doesn't run there). Everything else is green, including all mssql e2e jobs.

File: api/src/extensions/lib/sync/sync.ts [hunk 4]
```diff
@@ -69,39 +69,62 @@ export async function syncExtensions(options?: ExtensionSyncOptions): Promise<vo
 			}
 		}
 
-		// Make sure we don't overload the file handles
-		const queue = new Queue({ concurrency: 1000 });
+		// Limit concurrent requests to the storage driver (also bounds open file handles)
+		const queue = new Queue({ concurrency: Number(env['EXTENSIONS_SYNC_MAX_CONCURRENCY']) });
```
Comment (ComfortablyCoding) [inline api/src/extensions/lib/sync/sync.ts]: Ideally, this should be shared across all extension storage location "sync" operations (e.g. install). A more generic name, such as `EXTENSIONS_STORAGE_MAX_CONCURRENCY`, seems more appropriate.

Additionally, this should include validation and/or defaulting, pqueue will throw a type error if the value is not a number greater than or equal to 1.

File: .changeset/parallel-extension-sync.md [hunk 5]
```diff
@@ -0,0 +1,6 @@
+---
+'@directus/api': patch
+'@directus/env': patch
+---
+
+Fixed slow extension sync from remote storage on startup by downloading files concurrently instead of one at a time. The number of concurrent downloads is configurable via the new `EXTENSIONS_SYNC_MAX_CONCURRENCY` environment variable (default 20).
```
Comment (ComfortablyCoding) [inline .changeset/parallel-extension-sync.md]: Nit
```suggestion
Fixed slow extension sync from remote storage during startup. Added `EXTENSIONS_STORAGE_MAX_CONCURRENCY` to configure the maximum number of concurrent downloads.
```

**Review by ComfortablyCoding — CHANGES_REQUESTED** [review]

File: api/src/extensions/lib/sync/sync.ts (hunk 4 above)
Comment (dstockton) [reply api/src/extensions/lib/sync/sync.ts]: Done in 5aeadaa. Renamed to EXTENSIONS_STORAGE_MAX_CONCURRENCY and pulled it into a shared getStorageMaxConcurrency() helper that the sync queue and the marketplace install/uninstall storage queues all use (those two were hard-coded to 1000 as well). Invalid values (non-integer, < 1) now log a warning and fall back to the default 20 instead of hitting p-queue's TypeError. Unit tests cover the boundary cases and I verified the fallback end to end against minio. Docs PR updated for the rename: directus/docs#789

File: .changeset/parallel-extension-sync.md (hunk 5 above)
Comment (dstockton) [reply .changeset/parallel-extension-sync.md]: Applied, thanks.

**Review by ComfortablyCoding — APPROVED** [review]
LGTM!
